### PR TITLE
Cache functions by name.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: libsolidity/ast/AST.h [hunk 1]
```diff
@@ -496,7 +496,25 @@ class ContractDefinition: public Declaration, public StructurallyDocumented, pub
...
 	std::vector<ModifierDefinition const*> functionModifiers() const { return filteredNodes<ModifierDefinition>(m_subNodes); }
-	std::vector<FunctionDefinition const*> definedFunctions() const { return filteredNodes<FunctionDefinition>(m_subNodes); }
+	// TODO use range view or whatever
+	std::vector<FunctionDefinition const*> definedFunctions(std::string const* _name = nullptr) const {
+		vector<FunctionDefinition const*> ContractDefinition::definedFunctions(string const* _name) const
+		{
+			std::multimap<std::string, FunctionDefinition const*> const& functions = m_definedFunctionsByName.init([&]{
+				std::multimap<std::string, FunctionDefinition const*> result;
+				for (FunctionDefinition const* fun: filteredNodes<FunctionDefinition>(m_subNodes))
+					result.insert({fun->name(), fun});
```
Comment (chriseth) [inline libsolidity/ast/AST.h]: FunctionDefinition is only forward-declared here.

File: libsolidity/ast/AST.h [hunk 2]
```diff
@@ -496,7 +499,13 @@ class ContractDefinition: public Declaration, public StructurallyDocumented, pub
...
 	std::vector<VariableDeclaration const*> stateVariables() const { return filteredNodes<VariableDeclaration>(m_subNodes); }
 	std::vector<ModifierDefinition const*> functionModifiers() const { return filteredNodes<ModifierDefinition>(m_subNodes); }
-	std::vector<FunctionDefinition const*> definedFunctions() const { return filteredNodes<FunctionDefinition>(m_subNodes); }
+	auto definedFunctions() const {
+		return definedFunctionsByName() | ranges::views::transform([](auto const& _item) { return _item.second; });
+	}
+	auto definedFunctions(std::string const& _name) const {
+		auto&& [b, e] = definedFunctionsByName().equal_range(_name);
+		return ranges::span<decltype(*b)>(b, e) | ranges::views::transform([](auto const& _item) { return _item.second; });
+	}
```
Comment (leonardoalt) [inline libsolidity/ast/AST.h:509]: @ekpyron is that you

File: libsolidity/ast/AST.h [hunk 3]
```diff
@@ -496,7 +499,13 @@ class ContractDefinition: public Declaration, public StructurallyDocumented, pub
 	std::vector<EnumDefinition const*> definedEnums() const { return filteredNodes<EnumDefinition>(m_subNodes); }
 	std::vector<VariableDeclaration const*> stateVariables() const { return filteredNodes<VariableDeclaration>(m_subNodes); }
 	std::vector<ModifierDefinition const*> functionModifiers() const { return filteredNodes<ModifierDefinition>(m_subNodes); }
-	std::vector<FunctionDefinition const*> definedFunctions() const { return filteredNodes<FunctionDefinition>(m_subNodes); }
+	auto definedFunctions() const {
+		return definedFunctionsByName() | ranges::views::transform([](auto const& _item) { return _item.second; });
```
Comment (leonardoalt) [inline libsolidity/ast/AST.h]:
```suggestion
		return definedFunctionsByName() | ranges::views::values;
```
does that work?

File: libsolidity/ast/AST.h (hunk 2 above)
Comment (ekpyron) [reply libsolidity/ast/AST.h:509]: That's what we wrote about in the chat earlier :-).
I think it has to be
```
return ranges::subrange(b, e) | ranges::views::values;
```
to work.

File: libsolidity/ast/AST.h (hunk 3 above)
Comment (ekpyron) [reply libsolidity/ast/AST.h]: Yes, that should work, when including ``<range/v3/view/map.hpp>`` and is nicer.

Comment (chriseth) [reply libsolidity/ast/AST.h]: I tried it, but it doesn't - it's a multimap - d'oh!

Comment (chriseth) [reply libsolidity/ast/AST.h]: No really, I'm underwhelmed by the power of that library.

Comment (ekpyron) [reply libsolidity/ast/AST.h]: It does work for me for a multimap :-).

Comment (ekpyron) [reply libsolidity/ast/AST.h]: At least this:
```
#include <range/v3/all.hpp>
#include <iostream>
#include <map>

int main()
{

        std::multimap<int, int> x{{1,1},{2,3},{2,4}};
        auto r = x | ranges::views::values;
        for (auto v : r)
                std::cout << v << std::endl;
        std::cout << std::endl;

        auto&& [b,e] = x.equal_range(2);
        auto r2 = ranges::subrange(b,e) | ranges::views::values;
        for (auto v : r2)
                std::cout << v << std::endl;
}
```
compiles just fine for me.

Comment (chriseth) [reply libsolidity/ast/AST.h]: Weird...

**Comment by chriseth** [comment]
Tests fail because the order of iteration is different. I think it would be good to preserve it - I'll think about it.

File: libsolidity/ast/AST.cpp [hunk 4]
```diff
@@ -29,6 +29,9 @@
 #include <libsolidity/ast/TypeProvider.h>
 #include <libsolutil/Keccak256.h>
 
+#include <range/v3/view/map.hpp>
+#include <range/v3/span.hpp>
+
```
Comment (ekpyron) [inline libsolidity/ast/AST.cpp]: Neither are needed here anymore, are they?

File: libsolidity/ast/AST.h [hunk 5]
```diff
@@ -36,6 +36,10 @@
 
 #include <json/json.h>
 
+#include <range/v3/view/subrange.hpp>
+#include <range/v3/view/map.hpp>
+#include <range/v3/view/transform.hpp>
```
Comment (ekpyron) [inline libsolidity/ast/AST.h]: transform isn't needed anymore, is it?

**Review by ekpyron — COMMENTED** [review]
Apart from the iteration order question, this looks good to me apart from superfluous headers.

**Comment by ekpyron** [comment]
Regarding the order: Since C++11 the standard dictates ``The order of the key-value pairs whose keys compare equivalent is the order of insertion and does not change.``... with different keys the sorting is of course per name, though, but not sure that's too bad.

And I don't think there will be a way to have efficient lookup by name without sorting by name.

**Review by ekpyron — DISMISSED** [review]

**Review by ekpyron — DISMISSED** [review]
Needs rebase, apart from that still approved.

**Review by ekpyron — APPROVED** [review]

**Pull request merged**